src/cid_collections/hash_map.rs

```diff
@@ -0,0 +1,321 @@
+// Copyright 2019-2023 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use super::{CidV1DagCborBlake2b256, MaybeCompactedCid, Uncompactable};
+use cid::Cid;
+use std::collections::hash_map::{
+    Entry as StdEntry, IntoIter as StdIntoIter, OccupiedEntry as StdOccupiedEntry,
+    VacantEntry as StdVacantEntry,
+};
+
+/// A space-optimised hashmap of [`Cid`]s.
+///
+/// We accept the implementation complexity of per-compaction-method HashMaps for
+/// the space savings.
```

Comment:
This documentation is quite spartan. The previous docs elaborated more on our assumptions and expected space savings. Let's add documentation that is at least as detailed as it used to be.

Reply:
I think the documentation _is_ as detailed as it used to be, it's just moved to a common place.
It used to be repeated between e.g `CidHashMap`
https://github.com/ChainSafe/forest/blob/b0a2c2ca17edbff236810cacf3139d1123cf33ff/src/ipld/cid_hashmap.rs#L9-L19
and `SmallCid`
https://github.com/ChainSafe/forest/blob/b0a2c2ca17edbff236810cacf3139d1123cf33ff/src/utils/cid/mod.rs#L34-L46

Now there's _one_ place, and one struct with space - saving responsibility, and that's `MaybeCompactedCid`

I can add some design notes

Reply:
But the space savings for `CidHashMap` are quite different than the space savings for `SmallCid`. There are space savings for using `Vec<SmallCid>` vs. `Vec<Cid>,` and there are space savings for using `CidHashSet` vs. `HashSet<SmallCid>`. You are accidentally mixing these up and writing factually incorrect documentation (such as the ~20% space savings for `SmallCid`).